src/main/java/com/couchbase/lite/replicator/PusherInternal.java

```diff
@@ -98,14 +115,48 @@ public void setCreateTarget(boolean createTarget) {
     }
 
     @Override
+    protected void start() {
+        // create single thread supportExecutor for push relication
+        initSupportExecutor();
+        super.start();
+    }
+
+    // create single thread supportExecutor for push replication
+    private void initSupportExecutor() {
+        if (supportExecutor == null || supportExecutor.isShutdown()) {
+            supportExecutor = Executors.newSingleThreadExecutor(new ThreadFactory() {
+                @Override
+                public Thread newThread(Runnable r) {
+                    String maskedRemote = remote.toExternalForm();
+                    maskedRemote = maskedRemote.replaceAll("://.*:.*@", "://---:---@");
+                    return new Thread(r, "CBLPusherSupportExecutor-" + maskedRemote);
+                }
+            });
+        }
+    }
+
+    // shutdown supportExecutor immediately
+    private void terminateSupportExecutor() {
+        if (supportExecutor != null && !supportExecutor.isShutdown()) {
+            Utils.shutdownAndAwaitTermination(supportExecutor, 0, 5);
+        }
+    }
+
+    @Override
     protected void stop() {
         if (stateMachine.isInState(ReplicationState.STOPPED))
             return;
 
-        Log.d(Log.TAG_SYNC, "%s STOPPING...", toString());
+        Log.d(TAG, "%s STOPPING...", toString());
 
         stopObserving();
 
+        // Awake thread if it is wait for pause
+        setPaused(false);
+
+        // shutdown supportExecutor immediately, does not add any more tasks.
+        terminateSupportExecutor();
```

Comment:
Could this cause race condition between terminateSupportExecutor() and adding tasks to the executor? The race condition might be ok if adding the tasks to the work executor after being terminated has no side effect.